quickwit-storage/src/debouncer.rs

```diff
@@ -0,0 +1,307 @@
+/// data and the future. This is reflected on the generic type bounds for the value V: Clone.
+///
+/// Since most Futures return an Result<V, Error>, this also encompasses the error.
+pub struct AsyncDebouncer<K, V: Clone> {
+    cache: Mutex<FnvHashMap<K, Shared<BoxFuture<'static, V>>>>,
+}
+
+impl<K, V: Clone> Default for AsyncDebouncer<K, V> {
+    fn default() -> Self {
+        Self {
+            cache: Default::default(),
+        }
+    }
+}
+
+impl<K: Hash + Eq + Clone, V: Clone> AsyncDebouncer<K, V> {
+    /// Instead of the future directly, a constructor to build the future is passed.
+    /// In case there is already an existing Future for the passed key, the constructor is not
+    /// used.
+    pub async fn get_or_create<T, F>(&self, key: K, build_a_future: T) -> V
+    where
+        T: FnOnce() -> F,
+        F: Future<Output = V> + Send + 'static,
+    {
+        // explicit scope to drop the lock
+        let fut_opt = { self.cache.lock().unwrap().get(&key).cloned() };
+        if let Some(future) = fut_opt {
+            return future.await;
+        }
+
+        let fut = Box::pin(build_a_future()) as BoxFuture<'static, V>;
+        let fut = fut.shared();
+        self.cache.lock().unwrap().insert(key.clone(), fut.clone());
+        let res = fut.await;
+
+        self.cache.lock().unwrap().remove(&key);
+
+        res
+    }
+}
+
+type DebouncerKey = (PathBuf, Range<usize>);
+
+pub(crate) struct DebouncedStorage<T> {
+    // wrap both in Arc, because the Future is stored in the cache, which has 'static lifetime
+    // associated
+    underlying: Arc<T>,
+    slice_debouncer: Arc<AsyncDebouncer<DebouncerKey, StorageResult<OwnedBytes>>>,
+}
+
+impl<T: Storage> DebouncedStorage<T> {
+    pub(crate) fn new(underlying: T) -> Self {
+        Self {
+            underlying: Arc::new(underlying),
+            slice_debouncer: Arc::new(AsyncDebouncer::default()),
+        }
+    }
+}
+
+#[async_trait]
+impl<T: Storage> Storage for DebouncedStorage<T> {
+    async fn check(&self) -> anyhow::Result<()> {
+        self.underlying.check().await
+    }
+
+    async fn put(
+        &self,
+        path: &Path,
+        payload: Box<dyn crate::PutPayload>,
+    ) -> crate::StorageResult<()> {
+        self.underlying.put(path, payload).await
+    }
+
+    async fn copy_to_file(&self, path: &Path, output_path: &Path) -> StorageResult<()> {
+        self.underlying.copy_to_file(path, output_path).await
+    }
+
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> StorageResult<OwnedBytes> {
+        let (debouncer, underlying) = (self.slice_debouncer.clone(), self.underlying.clone());
+        let key = (path.to_owned(), range);
+        debouncer
+            .get_or_create(key.clone(), || async move {
+                underlying.get_slice(&key.0, key.1).await
+            })
+            .await
+    }
+
+    async fn delete(&self, path: &Path) -> StorageResult<()> {
+        self.underlying.delete(path).await
```

Comment:
I think there is a very complicated race conditions.

It makes i possible to have

- start get #1
- start delete #2
- end delete #2
- start get #3
- end get #2
- end get #3

Where get #3 returns a key that added by delete.

Reply:
My example does not consider delete to be debounced.

Just Get being debounced with the current code.